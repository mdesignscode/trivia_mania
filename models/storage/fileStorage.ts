#!/usr/bin/env node
import { readFileSync, writeFileSync } from "fs";
import Question from "../question";
import User from "../user";
import { IQuestion, IUser, IUserStats } from "../interfaces";

interface IFilters {
  difficulty?: string;
  categories?: Array<string>;
}

export type QuestionsRecord = Record<string, Record<string, Question>>;
type SerializedQuestionsRecord = Record<string, Record<string, IQuestion>>;

export interface IStorageObjects {
  Questions: QuestionsRecord;
  Users: Record<string, User>;
}

/**
 * Storage engine for storing objects in file storage
 * @date 04/09/2023 - 09:54:39
 *
 * @class FileStorage
 * @typedef {FileStorage}
 */
class FileStorage {
  private filePath: string = "file.json";
  private objects: IStorageObjects = {
    Questions: {},
    Users: {},
  };

  /**
   * Adds a question or a list of questions to storage
   * @date 05/09/2023 - 20:36:02
   *
   * @param {Question} questions - the questions to be added
   */
  newQuestion(questions: Question | Array<Question>): void {
    if (Array.isArray(questions)) {
      questions.forEach((obj) => {
        // category entry
        if (!this.objects.Questions[obj.category])
          this.objects.Questions[obj.category] = {
            [obj.id]: obj,
          };
        else this.objects.Questions[obj.category][obj.id] = obj;

        // difficulty entry
        if (!this.objects.Questions[obj.difficulty])
          this.objects.Questions[obj.difficulty] = {
            [obj.id]: obj,
          };
        else this.objects.Questions[obj.difficulty][obj.id] = obj;
      });
    } else {
      const obj = questions;
      // category entry
      if (!this.objects.Questions[obj.category])
        this.objects.Questions[obj.category] = {
          [obj.id]: obj,
        };
      else this.objects.Questions[obj.category][obj.id] = obj;

      // difficulty entry
      if (!this.objects.Questions[obj.difficulty])
        this.objects.Questions[obj.difficulty] = {
          [obj.id]: obj,
        };
      else this.objects.Questions[obj.difficulty][obj.id] = obj;
    }
  }

  /**
   * Retrieves all questions by filter
   * @date 29/09/2023 - 15:12:44
   *
   * @param {string} filter
   * @param {string} [userId=""] - filters out questions a user hasn't answered yet
   * @returns {Record<string, Question>}
   */
  getQuestionsByFilter(
    filter: string,
    userId: string = ""
  ): Record<string, Question> {
    const filteredQuestions = this.objects.Questions[filter];
    let uniqueQuestions: Record<string, Question> = {};
    const user = this.getUser(userId);

    if (user) {
      const userAnswered = user.answeredQuestions;
      for (const key in filteredQuestions) {
        if (!userAnswered.includes(key)) {
          uniqueQuestions[key] = filteredQuestions[key];
        }
      }
    }

    return userId ? uniqueQuestions : filteredQuestions;
  }

  /**
   * Retrieve a question of a specific filter from storage
   * @date 05/09/2023 - 20:51:30
   *
   * @param {string} filter - Either `difficulty name` or `category name`
   * @param {string} id - the id of the question
   * @returns {Question}
   */
  getQuestion(filter: string, id: string): Question {
    return this.objects.Questions[filter][id];
  }

  /**
   * Retrieves all questions in storage
   * @date 29/09/2023 - 12:38:53
   *
   * @param {boolean} [singleRecord=false] - if true, retrieves a unique set of questions
   * @param {string} [userId=""] - retrieves only questions a user hasn't answered yet
   * @returns {(Record<string, Question> | QuestionsRecord)}
   */
  getAllQuestions(
    singleRecord: boolean = false,
    userId: string = ""
  ): Record<string, Question> | QuestionsRecord {
    const allQuestions: QuestionsRecord = {};
    const uniqueQuestions: Record<string, Question> = {};
    const user = this.getUser(userId);

    const storageQuestions = this.objects.Questions;
    for (const filter in storageQuestions) {
      allQuestions[filter] = {};
      for (const question in storageQuestions[filter]) {
        const questionObj = storageQuestions[filter][question];
        if (user && !user.answeredQuestions.includes(questionObj.id)) {
          uniqueQuestions[question] = questionObj;
          allQuestions[filter][questionObj.id] = questionObj;
        } else if (!userId) {
          uniqueQuestions[question] = questionObj;
          allQuestions[filter][questionObj.id] = questionObj;
        }
      }
    }

    return singleRecord ? uniqueQuestions : allQuestions;
  }

  /**
   * Returns a list of filtered questions based on provided filters
   * @date 04/09/2023 - 13:11:21
   *
   * @param {IFilters} filters
   * @returns {Array<Question>}
   */
  filterQuestions(filters: IFilters): Array<Question> {
    const questions = this.getAllQuestions(true);
    const filteredCategories: Array<Question> = filters.categories?.length
      ? Object.values(questions).filter((question) =>
          filters.categories?.includes(question.category)
        )
      : Object.values(questions).map((question) => question);
    const filteredQuestions = filters.difficulty
      ? filteredCategories.filter(
          (question) => question.difficulty === filters.difficulty
        )
      : filteredCategories;

    // shuffle questions
    function shuffleArray(array: Array<Question>): Array<Question> {
      for (let i = array.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [array[i], array[j]] = [array[j], array[i]]; // Swap elements
      }
      return array;
    }

    return shuffleArray(filteredQuestions);
  }

  /**
   * counts all question filters, or counts all categories by difficulty
   * @date 29/09/2023 - 15:05:49
   *
   * @param {string} [difficulty=""]
   * @param {string} [userId=""] - optionally counts filters user has not answered
   * @returns {Record<string, number>}
   */
  questionsStats(
    difficulty: string = "",
    userId: string = ""
  ): Record<string, number> {
    const stats: Record<string, number> = {};
    let uniqueQuestions: Record<string, Question>;

    uniqueQuestions = difficulty
      ? this.getQuestionsByFilter(difficulty, userId)
      : (this.getAllQuestions(true, userId) as Record<string, Question>);

    const uniqueValues = Object.values(uniqueQuestions);
    if (difficulty) {
      for (const value of uniqueValues) {
        const category = value.category;
        if (!stats[category]) {
          const statCount = uniqueValues.filter((question) => {
            return question.category === category;
          });
          stats[category] = statCount.length;
        }
      }
      stats["all categories"] = Object.keys(uniqueQuestions).length;
    } else {
      const allQuestions = this.getAllQuestions(false, userId);
      for (const key in allQuestions) {
        stats[key] = Object.keys(allQuestions[key]).length;
      }

      stats["all difficulties"] = Object.keys(uniqueQuestions).length;
      stats["all categories"] = Object.keys(uniqueQuestions).length;
    }

    return stats;
  }

  /**
   * Adds a user or a list of users to storage
   * @date 15/09/2023 - 14:28:56
   *
   * @param {(User | Array<User>)} user
   */
  newUser(user: User | Array<User>) {
    if (Array.isArray(user)) {
      const users = user as Array<User>;

      users.forEach((user) => (this.objects.Users[user.id] = user));
    } else {
      const NewUser = user as User;
      this.objects.Users[NewUser.id] = user;
    }
  }

  /**
   * deletes a user from storage
   * @date 15/09/2023 - 13:36:47
   *
   * @param {string} id
   */
  deleteUser(id: string) {
    const users = this.objects.Users;

    if (users && users[id]) delete users[id];
  }

  /**
   * Retrieves a user
   * @date 11/09/2023 - 00:45:12
   *
   * @param {string} id - the user's id
   * @returns {User | null}
   */
  getUser(id: string): User | null {
    return this.objects.Users[id] || null;
  }

  /**
   * Saves a user's progress on a round
   * @date 11/09/2023 - 16:12:07
   *
   * @param {string} id - the user's id
   * @param {UserStats} stats
   */

  /**
   * Saves a user's progress on a round
   * @date 11/09/2023 - 16:12:07
   *
   * @param {string} id - the user's id
   * @param {IUserStats} stats - new stats to be added
   * @param {string[]} questionsAnswered - list of answered questions
   */
  updateUserProgress(id: string, stats: IUserStats, questionsAnswered: string[]) {
    const user = this.getUser(id);
    if (user) {
      user.submitRound(stats, questionsAnswered);
      this.save();
    }
  }

  /**
   * Retrieves a user's progress
   * @date 11/09/2023 - 00:51:09
   *
   * @param {string} id - the user's id
   * @returns {UserStats | undefined}
   */
  getUserStats(id: string): IUserStats | undefined {
    const user = this.getUser(id);
    if (user) {
      return user.stats;
    }
  }

  /**
   * Retrieves all users in storage
   * @date 11/09/2023 - 17:51:09
   *
   * @returns {Record<string, User>}
   */
  getAllUsers(): Record<string, User> {
    return this.objects.Users;
  }

  /**
   * Retrieves top ten scoring users
   * @date 11/09/2023 - 00:54:37
   *
   * @returns {Array<User>}
   */
  getTopTenUsers(): Array<User> {
    const usersList = Object.values(this.objects.Users) as Array<User>;
    const users = usersList.filter(
      (user) => Object.keys(user.stats).length > 0
    );
    const sortedUsers = users.sort((a, b) => {
      const userA = calculateTotalScore(a as User);
      const userB = calculateTotalScore(b as User);
      return userB - userA;
    });

    function calculateTotalScore(user: User): number {
      return user.stats.total.correctAnswered;
    }

    return sortedUsers.slice(0, 10) as Array<User>;
  }

  /**
   * Serializes all objects to storage file
   */
  save(): void | any {
    const jsonData = JSON.stringify(this.objects);
    try {
      writeFileSync(this.filePath, jsonData, "utf-8");
    } catch (error) {
      return error;
    }
  }

  /**
   * Deserializes objects in file storage
   */
  reload(): void {
    const data: IStorageObjects = {
      Users: {},
      Questions:{}
    };
    try {
      const parsedData: Record<string, Record<string, any>> = JSON.parse(
        readFileSync(this.filePath, "utf-8")
      );

      // recreate Question models
      for (const parsedKey in parsedData.Questions as SerializedQuestionsRecord) {
        const parsedObj = parsedData.Questions[parsedKey] as Record<string, IQuestion>
        data.Questions[parsedKey] = {}
        for (const key in parsedObj) {
          const model = new Question(parsedObj[key]);
          data.Questions[parsedKey][key] = model
        }
      }

      // recreate User models
      data.Users = {};
      for (const key in parsedData.Users as Record<string, IUser>) {
        const model = parsedData.Users[key] as IUser;
        const user = new User(
          model.username,
          model.id,
          model.stats,
          model.avatar,
          model.answeredQuestions
        );

        data.Users[key] = user;
      }
    } catch (error) {}
    Object.assign(this.objects, data);
  }

  /**
   * Clears objects in memory. For testing purposes only!
   * @date 06/09/2023 - 13:28:43
   */
  clearMemory(): void {
    this.objects = {
      Questions: {},
      Users: {},
    };
  }
}

export default FileStorage;
