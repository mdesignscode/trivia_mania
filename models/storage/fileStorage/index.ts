#!/usr/bin/env node
/* Provides abstraction to application */

import Question from "../../question";
import User from "../../user";
import QuestionStorage from "./question";

export type QuestionsRecord = Record<string, Record<string, Question>>;
export interface IStorageObjects {
  Questions: QuestionsRecord;
  Users: Record<string, User>;
}

/**
 * Storage engine for storing objects in file storage
 * @date 16/10/2023 - 21:20:12
 *
 * @class FileStorage
 * @type {FileStorage}
 * @extends {QuestionStorage}
 */
class FileStorage extends QuestionStorage {

}

export default FileStorage;
