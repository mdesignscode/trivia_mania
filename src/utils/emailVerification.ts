import { env } from '$env/dynamic/private';
import { EmailVerification } from '../models';
import nodemailer from 'nodemailer';

const sender = env.EMAIL_ADDRESS;
const pass = env.EMAIL_CODE;

// Configure the transporter
const transporter = nodemailer.createTransport({
        host: 'smtp.gmail.com', // e.g. smtp.gmail.com
        port: 587,
        secure: false, // true for 465, false for other ports
        auth: {
                user: sender,
                pass,
        },
});

// Send email function
export const sendEmail = async ({ email, username }: { username: string, email: string }) => {
        // generate email verification code
        const code = await generateEmailVerificationCode(email);

        // email body
        const textBody = `<p>Hi ${username},</p>
    <p>Thank you for signing up with <strong>Trivia Mania</strong>! To complete your registration, please use the verification code below:</p>
    <div class="code-box">${code}</div>
    <p>This code will expire in <strong>10 minutes</strong>. If you did not request this, you can safely ignore this email.</p>
    <p>Need help? <a href="mailto:${sender}">Contact our support team</a>.</p>
    <div class="footer">
      &copy; 2025 Trivia Mania. All rights reserved.
      <br>
      <a href="https://triviamania.vercel.app">https://triviamania.vercel.app</a>
    </div>`

        try {
                // send email
                await transporter.sendMail({
                        from: '"Trivia Mania" ejue1vs7n@mozmail.com', // sender address
                        to: email, // list of receivers
                        subject: 'Verify Your Email Address', // Subject line
                        html: textBody, // HTML body
                });

                return { status: 'success', data: code };
        } catch (error) {
                console.error('Error sending email:', error);
                return { status: 'fail', data: error.message };
        }
};


export async function generateEmailVerificationCode(email: string): Promise<string> {
        const code = Math.floor(100000 + Math.random() * 900000).toString(); // 6-digit code
        const expiresAt = new Date(Date.now() + 10 * 60 * 1000); // 10 minutes from now

        // existing verification should be updated
        const existingVerification = await EmailVerification.findOne({
                where: { email },
        });

        if (existingVerification) {
                existingVerification.setDataValue('code', code);
                existingVerification.setDataValue('expiresAt', expiresAt);
                await existingVerification.save();
                return code;
        }

        await EmailVerification.create({
                email,
                code,
                expiresAt,
                used: false,
        });

        return code;
}

export async function validateEmailVerificationCode(email: string, inputCode: string): Promise<{ success: boolean; message: string }> {
        const record = await EmailVerification.findOne({
                where: {
                        email,
                },
        });

        const failedResponse = { success: false, message: 'Invalid or expired verification code.' };

        if (record) {
                // if code hasn't expired yet
                const codeExpired = record.getDataValue('expiresAt') < new Date();
                if (codeExpired) return failedResponse;

                // is code correct
                const code = record.getDataValue('code');
                if (code !== inputCode) return failedResponse;

                // verified
                record.setDataValue('used', true)
                await record.save();

                return { success: true, message: 'Verification successful.' };
        }

        // record not found
        return failedResponse;
}

