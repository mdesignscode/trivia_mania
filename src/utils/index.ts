import { browser } from '$app/environment';

export function decodeHTMLEntities(text: string) {
        if (browser) {
                const element = document.createElement("div");
                element.innerHTML = text;
                return element.textContent || "";
        }
}

export const colorMap: { [key: string]: string } = {
        easy: "green",
        medium: "gold",
        hard: "red",
};

export function capitalize(str: string): string {
	if (!str) return '';
	return str.charAt(0).toUpperCase() + str.slice(1).toLowerCase();
}

