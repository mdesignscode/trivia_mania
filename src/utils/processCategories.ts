export type TQuestionData = {
        category: string;
        difficulty: string;
        count: number;
};
export type TAllCategories = Record<string, Record<string, TQuestionData>>;

export type TProcessResult = {
        allCategories: TAllCategories;
        allDifficulties: string[];
};

export function processCategories(
        categories: TQuestionData[],
        difficulties: string[]
): TProcessResult {
        const allCategories: TAllCategories = {};

        // map difficulties to categories
        categories.forEach(({ difficulty, count, category }) => {
                if (!allCategories[category]) allCategories[category] = {};
                allCategories[category][difficulty] = { difficulty, count, category };
        });

        // add "all difficulties" count for each category
        Object.keys(allCategories).forEach((key) => {
                const easy = allCategories[key].easy?.count || 0;
                const medium = allCategories[key].medium?.count || 0;
                const hard = allCategories[key].hard?.count || 0;
                const allCount = easy + medium + hard;

                allCategories[key]['all difficulties'] = {
                        difficulty: 'all difficulties',
                        count: allCount,
                        category: key
                };
        });

        // total "All Categories" aggregate
        const totalCategories: Record<string, TQuestionData> = {
                easy: { difficulty: 'easy', count: 0, category: 'All Categories' },
                medium: { difficulty: 'medium', count: 0, category: 'All Categories' },
                hard: { difficulty: 'hard', count: 0, category: 'All Categories' },
                'all difficulties': {
                        difficulty: 'all difficulties',
                        count: 0,
                        category: 'All Categories'
                }
        };

        Object.values(allCategories).forEach((catGroup) => {
                ['easy', 'medium', 'hard'].forEach((level) => {
                        if (catGroup[level]) {
                                totalCategories[level].count += catGroup[level].count;
                                totalCategories['all difficulties'].count += catGroup[level].count;
                        }
                });
        });

        allCategories['All Categories'] = totalCategories;

        return {
                allCategories,
                allDifficulties: [...difficulties]
        };
}

