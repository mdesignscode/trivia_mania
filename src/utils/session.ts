import { db } from '$lib/db';
import { encodeBase32LowerCaseNoPadding, encodeHexLowerCase } from '@oslojs/encoding';
import { sha256 } from '@oslojs/crypto/sha2';

export function generateSessionToken(): string {
        const bytes = new Uint8Array(20);
        crypto.getRandomValues(bytes);
        return encodeBase32LowerCaseNoPadding(bytes);
}

export function createSession(token: string, userId: number) {
        const sessionId = encodeHexLowerCase(sha256(new TextEncoder().encode(token)));
        const expiresAt = new Date(Date.now() + 1000 * 60 * 60 * 24 * 30); // 30 days

        db.prepare(
                'INSERT INTO session (id, user_id, expires_at) VALUES (?, ?, ?)'
        ).run(sessionId, userId, Math.floor(expiresAt.getTime() / 1000));
}

export function validateSessionToken(token: string) {
        const sessionId = encodeHexLowerCase(sha256(new TextEncoder().encode(token)));

        const row = db.prepare(`
    SELECT session.id, session.user_id, session.expires_at, user.username
    FROM session
    INNER JOIN user ON user.id = session.user_id
    WHERE session.id = ?
  `).get(sessionId);

        if (!row) {
                return { session: null, user: null };
        }

        const session = {
                id: row.id as string,
                userId: row.user_id as number,
                expiresAt: new Date(row.expires_at * 1000)
        };
        const user = {
                username: row.username as string
        };

        if (Date.now() >= session.expiresAt.getTime()) {
                db.prepare('DELETE FROM session WHERE id = ?').run(session.id);
                return { session: null, user: null };
        }

        return { session, user };
}

export function invalidateSession(sessionId: string) {
        db.prepare('DELETE FROM session WHERE id = ?').run(sessionId);
}

