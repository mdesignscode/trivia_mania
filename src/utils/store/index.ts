import globalStore from './global.svelte';
import { questionStore, playStore } from './playStore.svelte';

export {
        globalStore, playStore, questionStore,
}

