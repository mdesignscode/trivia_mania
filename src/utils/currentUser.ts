import { redirect, type Cookies } from "@sveltejs/kit";
import { validateSessionToken } from "./session";

// check if user is already logged in
export const isSignedIn = async (token: string | undefined, route: string) => {
        // don't redirect auth pages
        const isAuthRoute = ['/login', '/signup'].includes(route);
        if (!token && isAuthRoute) return;

        // user not logged in
        if (!token) throw redirect(302, '/login');

        const { user } = await validateSessionToken(token);
        // user signed in
        if (user) throw redirect(302, '/');

        // no user
        if (!user && isAuthRoute) return;
        else throw redirect(302, '/login');
}

export const getUser = async (cookies: Cookies) => {
        const token = cookies.get('session');
        if (!token) {
                throw redirect(302, '/login');
        }

        const { session, user } = await validateSessionToken(token);
        if (!session || !user) {
                throw redirect(302, '/login');
        }

        const userData = user.dataValues;
        delete userData.password;

        return {
                user: userData,
        };
}

