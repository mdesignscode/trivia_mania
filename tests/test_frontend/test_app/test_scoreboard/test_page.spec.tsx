import ScoreBoard from "@/app/scoreboard/page";
import storage from "@/models/index";
import { screen, render } from "@/utils/test_utils"
import { mockTopTenUsers } from "./test_components/test_scoreboard.spec";

// Mock the storage module
jest.mock('@/models/index', () => ({
  getTopTenUsers: jest.fn(() => mockTopTenUsers),
}));

describe("ScoreBoard component", () => {
  it("Should render <Board />", async () => {
    // wait for component to be rendered on server
    const Component = await ScoreBoard()

    // render component
    render(Component)

    // assert behaviour
    const container = await screen.findByTestId("scoreboard-container")
    expect(container).toBeInTheDocument()
    expect(storage.getTopTenUsers).toBeCalled()
  })
});
