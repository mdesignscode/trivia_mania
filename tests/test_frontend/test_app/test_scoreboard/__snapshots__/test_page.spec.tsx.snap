// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board component snapshot matches 1`] = `
<body>
  <div>
    <div
      class="p-4"
      data-testid="scoreboard-container"
    >
      <h1
        class="text-2xl font-semibold mb-4"
      >
        Top 10 Scoreboard
      </h1>
      <div
        class="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-4"
      >
        <div
          class="bg-secondary dark:bg-accent-200 max-h-72 overflow-y-auto p-4 rounded-lg shadow-md col gap-3"
          data-testid="user-0-stat"
        >
          <div
            class="flex items-center"
          >
            <div
              class="mr-4"
            >
              <img
                alt="Avatar for mock user"
                class="rounded-full"
                data-nimg="1"
                decoding="async"
                height="50"
                loading="lazy"
                src="/_next/image?url=%2Favatar.png&w=128&q=75"
                srcset="/_next/image?url=%2Favatar.png&w=64&q=75 1x, /_next/image?url=%2Favatar.png&w=128&q=75 2x"
                style="color: transparent;"
                width="50"
              />
            </div>
            <div>
              <p
                class="text-xl font-semibold"
              >
                1
                . 
                mock user
              </p>
              <p
                class="text-gray-500"
              >
                Total Correct Answers: 
                0
              </p>
            </div>
          </div>
          <div
            class="flex gap-2 flex-col"
          />
        </div>
        <div
          class="bg-secondary dark:bg-accent-200 max-h-72 overflow-y-auto p-4 rounded-lg shadow-md col gap-3"
          data-testid="user-1-stat"
        >
          <div
            class="flex items-center"
          >
            <div
              class="mr-4"
            >
              <img
                alt="Avatar for mock user"
                class="rounded-full"
                data-nimg="1"
                decoding="async"
                height="50"
                loading="lazy"
                src="/_next/image?url=%2Favatar.png&w=128&q=75"
                srcset="/_next/image?url=%2Favatar.png&w=64&q=75 1x, /_next/image?url=%2Favatar.png&w=128&q=75 2x"
                style="color: transparent;"
                width="50"
              />
            </div>
            <div>
              <p
                class="text-xl font-semibold"
              >
                2
                . 
                mock user
              </p>
              <p
                class="text-gray-500"
              >
                Total Correct Answers: 
                0
              </p>
            </div>
          </div>
          <div
            class="flex gap-2 flex-col"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ScoreBoard component snapshot matches 1`] = `
<body>
  <div>
    <div
      class="p-4"
      data-testid="scoreboard-container"
    >
      <h1
        class="text-2xl font-semibold mb-4"
      >
        Top 10 Scoreboard
      </h1>
      <div
        class="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-4"
      >
        <div
          class="bg-secondary dark:bg-accent-200 max-h-72 overflow-y-auto p-4 rounded-lg shadow-md col gap-3"
          data-testid="user-0-stat"
        >
          <div
            class="flex items-center"
          >
            <div
              class="mr-4"
            >
              <img
                alt="Avatar for mock user"
                class="rounded-full"
                data-nimg="1"
                decoding="async"
                height="50"
                loading="lazy"
                src="/_next/image?url=%2Favatar.png&w=128&q=75"
                srcset="/_next/image?url=%2Favatar.png&w=64&q=75 1x, /_next/image?url=%2Favatar.png&w=128&q=75 2x"
                style="color: transparent;"
                width="50"
              />
            </div>
            <div>
              <p
                class="text-xl font-semibold"
              >
                1
                . 
                mock user
              </p>
              <p
                class="text-gray-500"
              >
                Total Correct Answers: 
                0
              </p>
            </div>
          </div>
          <div
            class="flex gap-2 flex-col"
          />
        </div>
        <div
          class="bg-secondary dark:bg-accent-200 max-h-72 overflow-y-auto p-4 rounded-lg shadow-md col gap-3"
          data-testid="user-1-stat"
        >
          <div
            class="flex items-center"
          >
            <div
              class="mr-4"
            >
              <img
                alt="Avatar for mock user"
                class="rounded-full"
                data-nimg="1"
                decoding="async"
                height="50"
                loading="lazy"
                src="/_next/image?url=%2Favatar.png&w=128&q=75"
                srcset="/_next/image?url=%2Favatar.png&w=64&q=75 1x, /_next/image?url=%2Favatar.png&w=128&q=75 2x"
                style="color: transparent;"
                width="50"
              />
            </div>
            <div>
              <p
                class="text-xl font-semibold"
              >
                2
                . 
                mock user
              </p>
              <p
                class="text-gray-500"
              >
                Total Correct Answers: 
                0
              </p>
            </div>
          </div>
          <div
            class="flex gap-2 flex-col"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
